### PR TITLE
Client hierarchy rework

### LINKED ISSUE #29: Client hierarchy should allow expanding non-ellipsed objects & arrays

### LINKED ISSUE #51: Client: Hierarchy mode for simplified view of named objects only
possibly fully pre-expanded using advanced ellipsing

### LINKED ISSUE #66: refactor HierarchyView
https://github.com/tjjfvi/escad/blob/dev/packages/client/src/HierarchyView.tsx

### PR DESCRIPTION
**Pull request by tjjfvi**
Fixes #66 
Fixes #29 
Fixes #51

### CODE REVIEW

File: packages/client/src/HierarchyView/State.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+export function getState(stateMemo: StateMemo, path: HierarchyPath, str: string){
+  const key = path.flatMap(part => [
+    part.type,
+    part.type === "ObjectHierarchyPathPart"
+      ? part.key
+      : part.type === "ArrayHierarchyPathPart"
+        ? part.index
+        : part.location,
+  ]).concat(str).join("//")
```
Comment (vjjft) [inline packages/client/src/HierarchyView/State.ts:16]: escape `key` or handle using `JSON.stringify`

Comment (vjjft) [inline packages/client/src/HierarchyView/State.ts:16]: @vjjft add deferred issue re state representation

File: packages/client/src/HierarchyPane.tsx [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
+import { HierarchyToTreeEngine, HierarchyView } from "./HierarchyView"
 import { Pane } from "./Pane"
+import { httDefaultEngine } from "./HierarchyView/httDefaultEngine"
+import { Dropdown } from "./Dropdown"
+import { httSimpleEngine } from "./HierarchyView/httSimpleEngine"
 
 export const HierarchyPane = observer(() => {
+  const engine = useObservable<HierarchyToTreeEngine>(httDefaultEngine)
   const state = useContext(ClientState.Context)
   const hierarchy = state.hierarchy()
```
Comment (vjjft) [inline packages/client/src/HierarchyPane.tsx:13]: consider renaming `ClientState.hierarchy` to `renderedHierarchy`

**Review by vjjft — CHANGES_REQUESTED** [review]

File: packages/client/src/Dropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+export interface DropdownProps<T> {
+  options: Record<string, T>,
+  value: Observable<T>,
+}
+
+export const Dropdown = observer(<T, >({ options, value }: DropdownProps<T>) => {
+  const key = Object.entries(options).find(([, v]) => v === value())?.[0]
+  if(!key) throw new Error("Value passed to Dropdown not in options")
+  return <div className="Dropdown">
+    <select value={key} onChange={event => value(options[event.target.value])}>
```
Comment (vjjft) [inline packages/client/src/Dropdown.tsx:15]:
```suggestion
export const Dropdown = observer(<T, >({ options, value }: DropdownProps<T>) => {
  const currentValue = value()
  const currentKey = Object.keys(options).find(key => options[key] === currentValue)
  if(!currentKey) throw new Error("Value passed to Dropdown not in options")
  return <div className="Dropdown">
    <select value={currentKey} onChange={event => value(options[event.target.value])}>
```

File: packages/client/src/HierarchyView/NestableSpan.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import React, { useState } from "react"
+
+export interface NestableSpanProps {
+  className?: string,
+  onClick?: (event: React.MouseEvent) => void,
+  children: React.ReactNode,
+}
+
+/**
+ * Correctly handles hover & click logic when nested
```
Comment (vjjft) [inline packages/client/src/HierarchyView/NestableSpan.tsx:11]: provide specific description of 'correctly'

File: packages/client/src/HierarchyPane.tsx (hunk 2 above)
Comment (tjjfvi) [reply packages/client/src/HierarchyPane.tsx:13]: Not related to this pr, can be considered later

File: packages/client/src/HierarchyView/State.ts (hunk 1 above)
Comment (tjjfvi) [reply packages/client/src/HierarchyView/State.ts:16]: Going to leave as-is for now

**Pull request merged**